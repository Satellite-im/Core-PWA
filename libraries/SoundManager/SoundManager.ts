import { Howl } from 'howler'
import { Config } from '~/config'

// Keep this type in sync with Config.sounds
export enum Sounds {
  NEW_MESSAGE = 'newMessage',
  CALL = 'call',
  HANGUP = 'hangup',
  MUTE = 'mute',
  UNMUTE = 'unmute',
  DEAFEN = 'deafen',
  UNDEAFEN = 'undeafen',
  UPLOAD = 'upload',
  CONNECTED = 'connected',
}

/**
 * Class representing a SoundManager
 * @class SoundManager
 */
export default class SoundManager {
  sounds: Record<Sounds, Howl>
  /**
   * @constructs SoundManager
   */
  constructor(volume: number = 1.0) {
    this.sounds = {} as Record<Sounds, Howl>
    for (const [key, value] of Object.entries(Config.sounds) as [
      Sounds,
      string
    ][]) {
      this.sounds[key] = new Howl({
        src: [`${Config.ipfs.gateway}${value}`],
        loop: false,
        volume,
        html5: true,
      })
    }
  }

  /** @function
   * Check if a specific sound exists
   * @param sound Name of the sound to check
   */
  private existsSound(sound: Sounds) {
    if (!this.sounds[sound]) {
      throw new Error('Sound not found')
    }
  }

  /** @function
   * Plays a specified sound
   * @name play
   * @argument sound Name of the sound file to play.
   * @returns null
   */
  playSound(sound: Sounds) {
    this.existsSound(sound)

    this.sounds[sound].play()
  }

  /** @function
   * Stops a specified sound
   * @name stop
   * @argument sound Name of the sound file to play.
   * @returns null
   */
  stopSound(sound: Sounds) {
    this.existsSound(sound)

    this.sounds[sound].stop()
  }

  /** @function
   * Checks if the sound is currently playing
   * @name isPlaying
   * @argument sound Name of the sound file to play.
   * @returns boolean | error
   */
  isPlaying(sound: Sounds): boolean {
    this.existsSound(sound)

    return this.sounds[sound].playing()
  }

  changeLevels(volume: number) {
    for (const [key] of Object.entries(Config.sounds) as [Sounds, string][]) {
      this.sounds[key].volume(volume)
    }
  }
}
