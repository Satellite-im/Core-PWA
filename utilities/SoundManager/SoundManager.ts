import { Howl } from 'howler'
import { Config } from '~/config'

// Keep this type in sync with Config.sounds
export enum Sounds {
  NewMessage = 'newMessage',
  Call = 'call',
  Hangup = 'hangup',
  Mute = 'mute',
  Unmute = 'unmute',
  Deafen = 'deafen',
  Undeafen = 'undeafen',
  Upload = 'upload',
  Connected = 'connected',
}

/**
 * Class representing a Soundmanager
 * @class SoundManager
 */
export default class SoundManager {
  sounds: Record<Sounds, Howl>
  /**
   * @constructs SoundManager
   */
  constructor() {
    this.sounds = {} as Record<Sounds, Howl>
    for (const [key, value] of Object.entries(Config.sounds) as [
      Sounds,
      string
    ][]) {
      this.sounds[key] = new Howl({
        src: [`${Config.ipfs.browser}${value}`],
        loop: false,
        volume: 0.8,
        html5: true,
      })
    }
  }

  /** @function
   * Plays a specified sound
   * @name play
   * @argument sound Name of the sound file to play.
   * @returns null
   */
  playSound(sound: Sounds) {
    if (!this.sounds[sound]) {
      console.error('Sound not found')
      return
    }

    this.sounds[sound].play()
  }

  /** @function
   * Stops a specified sound
   * @name stop
   * @argument sound Name of the sound file to play.
   * @returns null
   */
  stopSound(sound: Sounds) {
    if (!this.sounds[sound]) {
      console.error('Sound not found')
      return
    }

    this.sounds[sound].stop()
  }

  /** @function
   * Checks if the sound is currently playing
   * @name isPlaying
   * @argument sound Name of the sound file to play.
   * @returns boolean
   */
  isPlaying(sound: Sounds): boolean {
    if (!this.sounds[sound]) {
      console.error('Sound not found')
      return false
    }

    return this.sounds[sound].playing()
  }
}
