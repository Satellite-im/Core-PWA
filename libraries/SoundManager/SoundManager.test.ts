import { Howler } from 'howler'
import SoundManager, { Sounds } from './SoundManager'

describe('init', () => {
  it('should pass', () => {
    expect(Sounds).toMatchSnapshot(`
    Object {
    "CALL": "call",
    "CONNECTED": "connected",
    "DEAFEN": "deafen",
    "HANGUP": "hangup",
    "MUTE": "mute",
    "NEW_MESSAGE": "newMessage",
    "UNDEAFEN": "undeafen",
    "UNMUTE": "unmute",
    "UPLOAD": "upload",
    }
        `)
  })
})

describe('Manage sounds', () => {
  let inst: any
  const volume: number = 1.0

  beforeEach(() => {
    inst = new SoundManager(volume)
  })

  test('sound exists', () => {
    const result: any = inst.existsSound(Sounds.CALL)
    expect(result).toMatchSnapshot()
  })

  test("sound doesn't exists", () => {
    try {
      inst.existsSound(null)
    } catch (error) {
      expect(error).toBeInstanceOf(Error)
      expect(error).toHaveProperty('message', 'Sound not found')
    }
  })

  test('sound plays', () => {
    window.HTMLMediaElement.prototype.load = () => {
      return Promise.resolve()
    }
    window.HTMLMediaElement.prototype.play = () => {
      return Promise.resolve()
    }
    window.HTMLMediaElement.prototype.pause = () => {
      return Promise.resolve()
    }
    const spy = jest.spyOn(inst.sounds[Sounds.CALL], 'play')
    const result: any = inst.playSound(Sounds.CALL)

    expect(spy).toHaveBeenCalled()
    expect(result).toMatchSnapshot()
  })

  test('sound stops', () => {
    const spy = jest.spyOn(inst.sounds[Sounds.CALL], 'stop')
    const result: any = inst.stopSounds([Sounds.CALL])

    expect(spy).toHaveBeenCalled()
    expect(result).toMatchSnapshot()
  })

  test('sound is playing', () => {
    const spy = jest.spyOn(inst.sounds[Sounds.CALL], 'playing')
    const result: any = inst.isPlaying(Sounds.CALL)

    expect(spy).toHaveBeenCalled()
    expect(result).toMatchSnapshot()
  })

  test('change volume level of sound', () => {
    const spy = jest.spyOn(inst.sounds[Sounds.CALL], 'volume')
    const result: any = inst.changeLevels(volume)

    expect(spy).toHaveBeenCalled()
    expect(result).toMatchSnapshot()
  })

  test('set mute sounds', () => {
    const spy = jest.spyOn(Howler, 'mute')
    inst.setMuteSounds(false)

    expect(spy).toHaveBeenCalled()
    expect(spy).toHaveBeenCalledWith(false)
  })
})
