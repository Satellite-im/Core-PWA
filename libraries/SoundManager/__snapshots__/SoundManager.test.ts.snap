// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Manage sounds change volume level of sound 1`] = `undefined`;

exports[`Manage sounds sound exists 1`] = `undefined`;

exports[`Manage sounds sound is playing 1`] = `false`;

exports[`Manage sounds sound plays 1`] = `undefined`;

exports[`Manage sounds sound stops 1`] = `undefined`;

exports[`init should pass 1`] = `
Object {
  "CALL": "call",
  "CONNECTED": "connected",
  "DEAFEN": "deafen",
  "HANGUP": "hangup",
  "MUTE": "mute",
  "NEW_MESSAGE": "newMessage",
  "UNDEAFEN": "undeafen",
  "UNMUTE": "unmute",
  "UPLOAD": "upload",
}
`;
